pkg/tinydtls: add DTLS sock API implementation

pokgak opened the pull request:
### Contribution description

This PR implements the DTLS sock API introduced in #11909 using tinyDTLS. It adds an example application dtls-sock that follows the same behavior as the existing dtls-echo example.

### Testing procedure

Run the example application using two nodes.

Start a DTLS server on one node and get the IP:
```
> dtlss start
> ifconfig
[...]
```

Send a message from the other node to the server:
```
> dtlsc <IP server> <message>
```

PSK credentials are used by default. To test with ECC credentials, uncomment following line in Makefile:
```
# CFLAGS += -DDTLS_ECC
```

Using ECC credentials is a bit iffy currently. It works but takes too long (>10s) to send and receive the data. Probably the same problem as #11859.

### Issues/PRs references

Depends on PR #11909.
Earlier discussion in #10897

rfuentess commented:
I can  start to review the PR  until  this weekend.  Is it OK ?

rfuentess commented:
By the way,  @pokgak can you put the link to the previous discussion ?

miri64 commented:
> I can start to review the PR until this weekend. Is it OK ?

#11909 needs to be reviewed and ACK'd first, so it still has some time (sorry, forgot to set the "State: waiting for other PR" label)

pokgak commented:
@rfuentess, it'd be great if you could review it, there are some places where I'm not really sure about the implementation but as @miri64 said we still have time. In the meantime, I've updated #10897 with a list to track the current status.

pokgak commented:
Added commits to update the implementation to use the latest API. There are not much changes to the implementation of the API in `sock_dtls.c` aside from function renaming and removal of `sock_dtls_init_server()`.

I think there will no changes anymore to the API proposed in #11909, only refining the example left. So it should be okay to start reviewing this PR @rfuentess .

rfuentess commented:
I finally got some free time. 

The interface seems to be elegant and clean. At simple glance I don't see anything that requires special attention. 

I'll try to get some physical boards on the following weeks to play with it. Otherwise, I'll recover my IoT-lab account for it.

miri64 commented:
I think with #11909 merged this can be rebased.

miri64 commented:
I guess it also needs some adaptation to the changes we discussed there.

miri64 commented on examples/dtls-sock/dtls-client.c:
```diff
@@ -0,0 +1,167 @@
+    }
+
+#ifdef DTLS_PSK
+    credman_credential_t credential = {
```
How about making `credential` a static const and moving to the global variables. This way, you can leave the `#ifdef` out of this function.

pokgak replied:
I'll change that.

miri64 commented on examples/dtls-sock/server_keys.h:
```diff
@@ -0,0 +1,60 @@
+#endif /* DTLS_PSK */
+
+#ifdef DTLS_ECC
+static const unsigned char server_ecdsa_priv_key[] = {
```
`static const` in a header will duplicate the memory for every inclusion of the header. Rather use `#define`s here, or define a `common.c` where you can return a pointer to these arrays with a function.

pokgak replied:
`#define`s might be too complicated for a multiline variable like this with the line continuation character. If I understands you correctly, with the `common.c` idea the user will have to write more code to just pass the certs.

How about doing it like how the wolfSSL PR does it [here][1]. Then in the .c file where the certs will be used, these declarations will be needed:

```C
extern const unsigned char server_cert[];
extern const unsigned long server_cert_len;
```

[1]: https://github.com/RIOT-OS/RIOT/pull/10308/files#diff-09ce13303248a398e36e3febb371129a

miri64 replied:
> `#define`s might be too complicated for a multiline variable like this with the line continuation character.

Why would it be more complicated? You just need to add `\`s.

```C
#define SERVER_ECDSA_PRIV_KEY { \
    0xa3, 0x8e, 0x19, 0xd9, 0x25, 0xd8, 0xf4, 0xf4, \
    0x96, 0x47, 0x3e, 0x4f, 0xd4, 0x6a, 0x12, 0xb2, \
    0xf5, 0xa1, 0xb5, 0x06, 0xa2, 0x42, 0x92, 0x3c, \
    0xfd, 0xf5, 0x4c, 0x6d, 0x01, 0x75, 0x89, 0x6f, \
}
```

> How about doing it like how the wolfSSL PR does it [here](https://github.com/RIOT-OS/RIOT/pull/10308/files#diff-09ce13303248a398e36e3febb371129a). Then in the .c file where the certs will be used, these declarations will be needed:
> 
> ```c
> extern const unsigned char server_cert[];
> extern const unsigned long server_cert_len;
> ```

As var as I understand `const` instead of `static const` comes with some memory caveats, especially when assigning the variable, but I would need to read up on that myself again.

pokgak replied:
> Why would it be more complicated? You just need to add `\`s.

That will need the user to change the file themselves.

> How about doing it like how the wolfSSL PR does it [here](https://github.com/RIOT-OS/RIOT/pull/10308/files#diff-09ce13303248a398e36e3febb371129a). Then in the .c file where the certs will be used, these declarations will be needed:
> 
> ```c
> extern const unsigned char server_cert[];
> extern const unsigned long server_cert_len;
> ```

This uses certificate file converted to hex in a .c file using `xxd -i`, so no more changes to the file are needed.

miri64 replied:
> > Why would it be more complicated? You just need to add `\`s.
> 
> That will need the user to change the file themselves.

The current approach does so as well.

> This uses certificate file converted to hex in a .c file using `xxd -i`, so no more changes to the file are needed.

This = "the approach here" or the WolfSSL approach?

miri64 replied:
See https://github.com/RIOT-OS/RIOT/pull/11943#issuecomment-534987491

miri64 commented on examples/dtls-sock/dtls-client.c:
```diff
@@ -0,0 +1,160 @@
+#define SOCK_DTLS_CLIENT_TAG (2)
+
+#ifdef DTLS_ECC
+static ecdsa_public_key_t other_pubkeys[] = {
```
This needs to be `static const` now (would make sense anyway, as I don't assume you change this).

pokgak replied:
Fixed. Also fixed in `dtls-server.c`.

miri64 commented on examples/dtls-sock/dtls-client.c:
```diff
@@ -0,0 +1,160 @@
+
+#else /* ifdef DTLS_PSK */
+static uint8_t psk_id_0[] = PSK_DEFAULT_IDENTITY;
+static uint8_t psk_key_0[] = PSK_DEFAULT_KEY;
```
Dito

pokgak replied:
Fixed.

pokgak replied:
Sidenote: this will introduce error when compiling:
```
error: initialization discards 'const' qualifier from pointer target type
```

credman uses `void *` so `const void *` to `void *` gives this error. I learned just now that the opposite is okay though (`void *` to `const void *`). I'll make a PR to change this in credman.

miri64 replied:
Does this PR exist by now?

pokgak replied:
It is now #12299 :)

miri64 commented on examples/dtls-sock/dtls-client.c:
```diff
@@ -0,0 +1,160 @@
+                .y = client_ecdsa_pub_key_y,
+            },
+            .client_keys = other_pubkeys,
+            .client_keys_size = sizeof(other_pubkeys) / sizeof(other_pubkeys[0]),
```
Please use `ARRAY_sIZE` from `core/include/kernel_defines.h` for this.

pokgak replied:
Changed to `ARRAY_SIZE(other_pubkeys)`

miri64 commented:
Please rebase, #12299 was merged.

miri64 commented:
As 894ffe5 still does exist, I believe you did not rebase to current master ;-)

pokgak commented:
@rfuentess update on the review? I'm looking into constant-time when copying the credentials in credential callbacks (`get_psk()`, `get_ecc()`). So that part may still change.

pokgak commented:
> As [894ffe5](https://github.com/RIOT-OS/RIOT/commit/894ffe52a1ebe71d9892a3392d886a0dad8c120d) still does exist, I believe you did not rebase to current master ;-)

No, that is not a workaround for the error stated in [comment][1]. After #12299 merged, compiling the example gives following error:

```
error: assignment discards 'const' qualifier from pointer target type
```
 894ffe5 is needed to fix it. And I checked git log just in case and #12299 is in the history.

[1]: https://github.com/RIOT-OS/RIOT/pull/11943#discussion_r319457043

rfuentess commented:
> @rfuentess update on the review? I'm looking into constant-time when copying the credentials in credential callbacks (get_psk(), get_ecc()). So that part may still change.

Feel free to proceed :)    Sadly, I got super busy with my current work. I'll try to review again  on this weekend and  after the rebase is done.

miri64 commented:
> No, that is not a workaround for the error stated in [comment](https://github.com/RIOT-OS/RIOT/pull/11943#discussion_r319457043). After #12299 merged, compiling the example gives following error:
> 
> ```
> error: assignment discards 'const' qualifier from pointer target type
> ```
> 
> [894ffe5](https://github.com/RIOT-OS/RIOT/commit/894ffe52a1ebe71d9892a3392d886a0dad8c120d) is needed to fix it. And I checked git log just in case and #12299 is in the history.

Ah sorry, I got confused, because the PR number was mentioned in the commit message.

pokgak commented:
@rfuentess I'm new to security stuff and am learning as I go so some of this might be trivial to you :)

> I'm looking into constant-time when copying the credentials in credential callbacks (`get_psk()`, `get_ecc()`). So that part may still change.

I looked into the constant-time implementation and decided that it is not required here. For `verify_ecdsa_key()`, the public keys are well... public. So there's no secret needed to be protected here. As for constant-time copying of the keys (e.g. in `get_psk_info()`), there is a mention on that in [this BearSSL post][1], but it seems too complicated and might not be needed here.

So TLDR: no changes are done to the current code and it is ready for the next round of review. I already rebased this PR to the latest master. Should I squash this PR so that it will be easier to review?

[1]: https://www.bearssl.org/constanttime.html

rfuentess commented:
> So TLDR: no changes are done to the current code and it is ready for the next round of review. I already rebased this PR to the latest master. Should I squash this PR so that it will be easier to review?

To me is OK to make the rebase now. Just wait for confirmation from @miri64

miri64 commented:
> > So TLDR: no changes are done to the current code and it is ready for the next round of review. I already rebased this PR to the latest master. Should I squash this PR so that it will be easier to review?
> 
> To me is OK to make the rebase now. Just wait for confirmation from @miri64

As I don't know much about neither tinydtls or DTLS in general, I mostly want to act as a guide for this PR. I can give the final ACK (if the reviewer doesn't have push rights), and if nobody else wants, also can review in detail. But I rather prefer this PR reviewed by someone deeper in to the topic.

miri64 commented:
(so you may squash @pokgak. My comments were addressed for the most part. I guess https://github.com/RIOT-OS/RIOT/pull/11943#discussion_r319184444 is okay, as it is only used by the application)

pokgak commented:
Squashed. 

I actually want to change how the keys are included to [#11943 (comment)](https://github.com/RIOT-OS/RIOT/pull/11943#discussion_r319437011) and try generating a new key with OpenSSL and use `xxd -i` to convert it to hex format (and maybe document it in the `net/dtls` Doxygen) page but it is not as straightforward as I thought it would be. Main reason to do this would be so that we show *a* way (not multiple ways depending on which dtls example is used) of embedding the keys in RIOT. I might tackle this in another PR. For now this PR is ready for review.

rfuentess commented:
> I actually want to change how the keys are included to #11943 (comment) and try generating a new key with OpenSSL and use xxd -i to convert it to hex format (and maybe document it in the net/dtls Doxygen) page but it is not as straightforward as I thought it would be so I left it how it is for now.

For attempting to generate new keys I'll leave them outside this PR.  Also,  maybe you were trying to generate certificates with OpenSSL instead of  PSK or ECDSA (My last attempt one year ago was also fruitless).  

By the way, certificates are being intentionally dropped  by tinyDTLS  due to the highly memory resource requirement for it (if you think ECDSA is bad, wait for that one :P ). 

I'll try to review the PR  and test it on IoT-lab  this weekend :)

rfuentess commented:
Testing with m3 nodes.
- Client PSK with Server PSK  is **OK**
- Client ECC with Server ECC : is **OK** (super slow, but is expected)
- Client PSK with Server ECC (must fail)  **Kernel panic**
- Client ECC with Server PSK (must fail) ** Kernel panic**

**NOTE**: This is not the case with `dtls-echo`
----

Client PSK  with a server ECC  we have  a kernel panic.

```�main(): This is RIOT! (Version: 2019.10-devel-1025-g24226-sock_dtls_impl)
DTLS sock example application
All up, running the shell now
> dtlsc fe80::1711:6b10:65f8:8f32 "To fail"
dtlsc fe80::1711:6b10:65f8:8f32 "To fail"
0x800f25b
*** RIOT kernel panic:
FAILED ASSERTION.

*** halted.
```

**NOTE**: Server is Ok after this one.

---

For the other kernel case, I had the debug enabled

```> dtlsc fe80::1711:6b10:65fd:bd36
dtlsc fe80::1711:6b10:65fd:bd36
usage dtlsc <addr> <data>
> dtlsc fe80::1711:6b10:65fd:bd36 "To fail also"
dtlsc fe80::1711:6b10:65fd:bd36 "To fail also"
Jan 01 00:09:39 DEBG DTLSv12: initialize HASH_SHA256
Jan 01 00:09:39 DEBG send handshake packet of type: client_hello (1)
Jan 01 00:09:39 DEBG send header: (zu bytes):
00000000 16 FE FF 00 00 00 00 00  00 00 00 00 52 
Jan 01 00:09:39 DEBG send unencrypted: (zu bytes):
00000000 01 00 00 46 00 00 00 00  00 00 00 46 
Jan 01 00:09:39 DEBG send unencrypted: (zu bytes):
00000000 FE FD 00 00 02 43 EB F7  15 4F C2 0D F9 2E 34 57 
00000010 E2 80 14 0A F2 17 1A 9B  24 F7 BD D0 EF CB 9F 50 
00000020 58 A7 00 00 00 02 C0 AE  01 00 00 1A 00 13 00 02 
00000030 01 02 00 14 00 02 01 02  00 0A 00 04 00 02 00 17 
00000040 00 0B 00 02 01 00 
Jan 01 00:09:39 DEBG dtls_handle_message: FOUND PEER
Jan 01 00:09:39 DEBG got packet 22 (44 bytes)
Jan 01 00:09:39 DEBG receive header: (zu bytes):
00000000 16 FE FF 00 00 00 00 00  00 00 00 00 1F 
Jan 01 00:09:39 DEBG receive unencrypted: (zu bytes):
00000000 03 00 00 13 00 00 00 00  00 00 00 13 FE FD 10 8A 
00000010 8E 51 7F 8E 37 9A 43 B2  CC 64 DC F2 7E 5C 3E 
Jan 01 00:09:39 DEBG received handshake packet of type: hello_verify_request (3)
Jan 01 00:09:39 DEBG handle handshake packet of type: hello_verify_request (3)
Jan 01 00:09:39 DEBG clear MAC
Jan 01 00:09:39 DEBG add MAC data: (zu bytes): 010000560001000000000056
Jan 01 00:09:39 DEBG add MAC data: (zu bytes): FEFD00000243EBF7154FC20DF92E3457E280140AF2171A9B24F7BDD0EFCB9F5058A700108A8E517F8E379A43B2CC64DCF27E5C3E0002C0AE0100001A001300020102001400020102000A000400020017000B00020100
Jan 01 00:09:39 DEBG send handshake packet of type: client_hello (1)
Jan 01 00:09:39 DEBG send header: (zu bytes):
00000000 16 FE FF 00 00 00 00 00  00 00 01 00 62 
Jan 01 00:09:39 DEBG send unencrypted: (zu bytes):
00000000 01 00 00 56 00 01 00 00  00 00 00 56 
Jan 01 00:09:39 DEBG send unencrypted: (zu bytes):
00000000 FE FD 00 00 02 43 EB F7  15 4F C2 0D F9 2E 34 57 
00000010 E2 80 14 0A F2 17 1A 9B  24 F7 BD D0 EF CB 9F 50 
00000020 58 A7 00 10 8A 8E 51 7F  8E 37 9A 43 B2 CC 64 DC 
00000030 F2 7E 5C 3E 00 02 C0 AE  01 00 00 1A 00 13 00 02 
00000040 01 02 00 14 00 02 01 02  00 0A 00 04 00 02 00 17 
00000050 00 0B 00 02 01 00 
Jan 01 00:09:39 DEBG dtls_handle_message: FOUND PEER
Jan 01 00:09:39 DEBG got packet 21 (15 bytes)
Jan 01 00:09:39 DEBG new packet arrived with seq_nr: lu
Jan 01 00:09:39 DEBG new bitfield is               : lx
Jan 01 00:09:39 DEBG receive header: (zu bytes):
00000000 15 FE FD 00 00 00 00 00  00 00 01 00 02 
Jan 01 00:09:39 DEBG receive unencrypted: (zu bytes):
00000000 02 28 
Jan 01 00:09:39 INFO ** Alert: level 2, description 40
Jan 01 00:09:39 ALRT 40 invalidate peer
Jan 01 00:09:39 WARN received alert, peer has been invalidated

0x801052f
*** RIOT kernel panic:
FAILED ASSERTION.

*** halted.
```

rfuentess commented:
Ok,  now I'm sure. With the ECC , the client is getting frozen.  Using a dtls-echo client with ECC  the performance is much different. 

So, we have problem with DTLS Alter records, and  maybe with DTLS data app after a timer expired ?

pokgak commented:
> * Client PSK with Server ECC (must fail)  **Kernel panic**
> 
> * Client ECC with Server PSK (must fail) **Kernel panic**

Nice catch. I forgot to test these cases. BTW do you know what is the behavior of tinydtls if both PSK and ECC is enabled?

> **NOTE**: I'm not sure, but it's seems there is another issue with ECC. The DTLS handshake is slow but stable, but the client seems to get freeze after sending the DTLS Data App record

> Ok, now I'm sure. With the ECC , the client is getting frozen. Using a dtls-echo client with ECC the performance is much different.
> 
> So, we have problem with DTLS Alter records, and maybe with DTLS data app after a timer expired ?

I'll look into this and report back later today.

rfuentess commented:
> BTW do you know what is the behavior of tinydtls if both PSK and ECC is enabled?

*TLR; DR*: Depends on the DTLS server, normally it would pick the strongest (on security terms). Also, most of the nodes doesn't have enough memory space for them.
 
To be honest when I'm speaking about PSK and ECC what I'm really speaking are Cipher suites (But tinyDTLs only has one of each type). So, if a client supports both of them, he will send a list of the cipher suites to the server, and this one will pick the one its support (and if supports both, the one with best "metric").

In memory requirement,  If I remember well,  from the IoT-lab node list, only the M3 can compile with both of them.

pokgak commented:
I think there's a bug in the dtls-echo example. The server does not returns a hanshake failure when no matching cipher found.

According to the Section 7.4.1.2 of TLS 1.2 RFC, the server should return a handshake failure if no matching ciphersuite is found, which is what the example in this PR did:

```
   The cipher suite list, passed from the client to the server in the
   ClientHello message, contains the combinations of cryptographic
   algorithms supported by the client in order of the client's
   preference (favorite choice first).  Each cipher suite defines a key
   exchange algorithm, a bulk encryption algorithm (including secret key
   length), a MAC algorithm, and a PRF.  The server will select a cipher
   suite or, if no acceptable choices are presented, return a handshake
   failure alert and close the connection.  If the list contains cipher
   suites the server does not recognize, support, or wish to use, the
   server MUST ignore those cipher suites, and process the remaining
   ones as usual.
```

Regarding the kernel panic, it is caused by the client that keeps on using the server peer object after the peer is destroyed. This is fixed in 581b10a.

That said, with that commit applied I found a new problem. Test procedure:

1. Start with client ECC server PSK
2. Try sending to server. This will fail because no matching cipher found. Server then sends handshake failure alert to client
3. Reflash client, this time using PSK and try sending to server. This will also fail. The client will timeout. Server gave warning:

```
WARN The message sequence number is too small, expected 1, got: 0
```

I think this is a bug in tinydtls upstream. According to the RFC snippet above, the server should (a) return a handshake failure alert and (b) close the connection. (b) is not done, causing the server to still use the peer from the old handshake. I tried playing around with tinydtls code but haven't found a fix for this yet.

If you could confirm that this indeed a bug in tinydtls, that'll be great. I'll try to reproduce this with minimal tinydtls setup :)

pokgak commented:
> If you could confirm that this indeed a bug in tinydtls, that'll be great. I'll try to reproduce this with minimal tinydtls setup :)

Okay, I figured out how to test tinydtls on Linux and cannot reproduce this problem. So maybe not upstream tinydtls bug.

rfuentess commented:
> I think there's a bug in the dtls-echo example. The server does not returns a hanshake failure when no matching cipher found.
> ...
> Okay, I figured out how to test tinydtls on Linux and cannot reproduce this problem. So maybe not upstream tinydtls bug.

Ok, I'm a little lost here. Can you confirm  if  the first is still an issue to confirm ? (I'll try to review this on the night)

pokgak commented:
> I think there's a bug in the dtls-echo example. The server does not returns a hanshake failure when no matching cipher found.

This still need to be confirmed in the dtls-echo example, but not related to this PR.

> That said, with that commit applied I found a new problem. Test procedure:
> 
>     1. Start with client ECC server PSK
> 
>     2. Try sending to server. This will fail because no matching cipher found. Server then sends handshake failure alert to client
> 
>     3. Reflash client, this time using PSK and try sending to server. This will also fail. The client will timeout. Server gave warning:
> 
> 
> ```
> WARN The message sequence number is too small, expected 1, got: 0
> ```

This is the issue I found in this PR after fixing the kernel panic issue (581b10a) you mentioned before.

pokgak commented:
> Using `TINYDTLS_LOG=6` I can see the DTLS Alert record on the dtls-echo client (PSK).

Yes, the DTLS alert arrived but it is the wrong alert (close_notify instead of handshake_failure). More on that in this issue #12351.

> By the way, I'm not going to have enough time this week to test further this PR.

Noted. You can review it when you have time.

pokgak commented:
Pushed some commits.

> Ok, now I'm sure. With the ECC, the client is getting frozen. Using a dtls-echo client with ECC the performance is much different.

This is now fixed. The timeout used in dtls-server is too small and because of that the server missed the decrypted message from the client and did not echo back. This causes the client to keeps on waiting for an echo from the server indefinitely.

Test result from my testing of dtls-sock against dtls-client

```
sock = dtls-sock
echo = dtls-echo

  -------- -------------- -------- -------------- -------------------------------------------------------
  client   cipher suite   server   cipher suite   result
                                                  -------------------------------------------------------
                                                  native
  -------- -------------- -------- -------------- -------------------------------------------------------
  sock     ecc            sock     ecc            OK
  sock     psk            sock     psk            OK
  sock     ecc            sock     psk            client got handshake_failure (expected)
  sock     psk            sock     ecc            client got handshake_failure (expected)

  echo     ecc            sock     ecc            OK
  echo     psk            sock     psk            OK
  sock     ecc            echo     ecc            OK
  sock     psk            echo     psk            OK

  echo     ecc            sock     psk            client hangs (cannot handle handshake_failure) \*1
  echo     psk            sock     ecc            client hangs (cannot handle handshake_failure) \*1                          
  sock     ecc            echo     psk            client timeout and sent alert close_notify \*2
  sock     psk            echo     ecc            client got handshake_failure (expected)
  -------- -------------- -------- -------------- -------------------------------------------------------
\*1: issue in dtls-echo implementation, client does not handle fatal alert
\*2: default DTLS_MAX_BUF too small and causes the server to drop incoming client_hello from client
```

rfuentess commented:
I'll try to review with calm your commits this weekend.

miri64 commented on examples/dtls-sock/Makefile:
```diff
@@ -0,0 +1,67 @@
+RIOTBASE ?= $(CURDIR)/../..
+
+# TinyDTLS only has support for 32-bit architectures ATM
+BOARD_BLACKLIST := arduino-duemilanove arduino-mega2560 arduino-uno chronos \
```
Please use `FEATURES_REQUIRED += arch_32bit` as a dependency for `tinydtls` (this is new ;-)).

pokgak replied:
Replaced `BOARD_BLACKLIST` with `FEATURES_REQUIRED`. Got warning about features not supported but rebasing to latest master solved it. Should I also move `BOARD_INSUFFICIENT_MEMORY` to `Makefile.ci ` like in #12406?

miri64 replied:
Yeah, otherwise the CI will now complain ;-).

pokgak replied:
Done.

miri64 commented on pkg/tinydtls/contrib/sock_dtls.c:
```diff
@@ -0,0 +1,474 @@
+        if ((timeout != SOCK_NO_TIMEOUT) && (timeout != 0)) {
+            timeout = timeout - (xtimer_now_usec() - start_recv);
+            DEBUG("sock_dtls: timeout left: %"PRIu32"\n", timeout);
+        }
```
Just to be clear: calculating the remaining timeout is just for debugging purposes, right? Maybe put a `#if ENABLE_DEBUG` around the `if` then?

smlng replied:
its used in L433 below again, but only compared for `0`. A possible problem here is that it rolls over due to subtraction - but I'm not sure if that is a problem at all?!

pokgak replied:
No, calculating the remaining timeout is not just for debugging purposes. `sock_udp_recv()` is called in a loop until a DTLS packet with application data is received and each time calling it the timeout will need to be recalculated. I'll admit that with how the timeout is implemented currently, it may not be enough for applications that require high-level of timing precision.

Regarding rollover, would a check like `timeout = (timeout < 0) ? 0 : timeout` okay to prevent the rollover?

smlng replied:
no because `timeout` is unsigned, hence cannot be `< 0`. you need to checkout if `(xtimer_now_usec() - start_recv)` is larger or smaller than `timeout` if it is larger, set timeout to 0

pokgak replied:
pushed a fix for this https://github.com/RIOT-OS/RIOT/pull/11943/commits/c6977c29a7de16b1693226f7eb2d96b6b7b49f0c

smlng commented on pkg/tinydtls/contrib/sock_dtls.c:
```diff
@@ -0,0 +1,474 @@
+int sock_dtls_create(sock_dtls_t *sock, sock_udp_t *udp_sock,
+                     credman_tag_t tag, unsigned version, unsigned role)
+{
+    assert(sock && udp_sock);
```
please use single conditions for `assert`, applies below too

pokgak replied:
fixed.

pokgak commented on examples/dtls-sock/dtls-server.c:
```diff
@@ -0,0 +1,207 @@
+    if (_dtls_server_pid == EOVERFLOW) {
+        puts("ERROR: Thread overflow!");
+        _dtls_server_pid = KERNEL_PID_UNDEF;
+        return;
```
Would combining it to one check and printing the error code be better? 

```C
if (_dtls_server_pid < 0) {
        printf("ERROR: failed to create thread: %d\n", _dtls_server_pid);
        _dtls_server_pid = KERNEL_PID_UNDEF;
}
```

smlng replied:
yes, to user it is all the same: thread creation failed -> server cannot be run. Hence, the exact failure might only be needed for debugging but that can be inferred from the return code.

TL;DR: change as you suggested (and remove/avoid unnecessary `return`)

pokgak replied:
Changed as suggested.

pokgak commented on pkg/tinydtls/contrib/sock_dtls.c:
```diff
@@ -0,0 +1,474 @@
+    remote.family = AF_INET6;
+
+    ssize_t res = sock_udp_send(sock->udp_sock, buf, len, &remote);
+    if (res <= 0) {
```
I can understand that `sock_udp_recv()` returning 0 is not an error but for `sock_udp_send()` if we failed to send any bytes, that means we failed to send the packet, right? Documentation for `sock_udp_send()` does not mention about return value 0.

smlng replied:
well what if I want to send an empty payload with len=0, this is possible for sock_udp_send (see docu on data)

pokgak replied:
Okay my bad, I misunderstood data as the whole packet and not just the payload.

pokgak commented on examples/dtls-sock/Makefile:
```diff
@@ -0,0 +1,57 @@
+# Comment this out to disable code in RIOT that does safety checking
+# which is not needed in a production environment but helps in the
+# development process:
+CFLAGS += -DDEVELHELP
```
Changed.

pokgak commented on examples/dtls-sock/Makefile:
```diff
@@ -0,0 +1,57 @@
+DTLS_PORT ?= 20220
+CFLAGS += -DDTLS_DEFAULT_PORT=$(DTLS_PORT)
+
+# NOTE: If not cipher suite is selected, DTLS_PSK is used by default.
```
fixed

pokgak commented on examples/dtls-sock/Makefile.ci:
```diff
@@ -0,0 +1,29 @@
+    spark-core \
+    stm32f0discovery \
+    stm32mindev \
+    yunjia-nrf51822
```
Added `#`.

pokgak commented on examples/dtls-sock/dtls-server.c:
```diff
@@ -0,0 +1,207 @@
+                           SOCK_DTLS_1_2, SOCK_DTLS_SERVER);
+    if (res < 0) {
+        puts("Error creating DTLS sock");
+        return 0;
```
Changed to `NULL`

pokgak commented on examples/dtls-sock/dtls-server.c:
```diff
@@ -0,0 +1,207 @@
+        return;
+    }
+
+    return;
```
Removed.

pokgak commented on examples/dtls-sock/dtls-server.c:
```diff
@@ -0,0 +1,207 @@
+        return;
+    }
+
+    if (_dtls_server_pid == EOVERFLOW) {
```
see comment above

pokgak commented on examples/dtls-sock/dtls-server.c:
```diff
@@ -0,0 +1,207 @@
+    if (_dtls_server_pid == EINVAL) {
+        puts("ERROR: Thread invalid");
+        _dtls_server_pid = KERNEL_PID_UNDEF;
+        return;
```
see comment above

pokgak commented on examples/dtls-sock/dtls-server.c:
```diff
@@ -0,0 +1,207 @@
+        stop_server();
+    }
+    else {
+        printf("Error: invalid command. Usage: %s start | stop\n", argv[0]);
```
Yup. You're right. Fixed.

pokgak commented on pkg/tinydtls/contrib/sock_dtls.c:
```diff
@@ -0,0 +1,474 @@
+}
+
+static int _event(struct dtls_context_t *ctx, session_t *session,
+           dtls_alert_level_t level, unsigned short code)
```
Fixed.

smlng commented on pkg/tinydtls/contrib/sock_dtls.c:
```diff
@@ -0,0 +1,474 @@
+    case DTLS_PSK_HINT:
+        DEBUG("sock_dtls: psk hint request\n");
+        /*
+        c = credential.params.psk.hint.s;
```
why is this commented out, either remove or use - or add proper documentation when this would be used and why it is currently not.

pokgak replied:
Added comment pointing to [RFC 4279 Section 5.2](https://tools.ietf.org/html/rfc4279#section-5.2) on why hint is ignored.

pokgak commented on pkg/tinydtls/Makefile.include:
```diff
@@ -27,6 +30,13 @@ ifneq (,$(filter tinydtls,$(USEMODULE)))
     endif
   endif
 
+  # Enable debug when using tinydtls_sock_dtls
```
This is not enabled by default. My goal here is to have a general debug flag `DTLS_DEBUG` or maybe `SOCK_DTLS_DEBUG` that can be enabled, independent of which DTLS stack are used to enable debugging when using DTLS sock. This flag, however, should not affect the existing way of enabling debugging for each DTLS stack (e.g. tinydtls uses `TINYDTLS_DEBUG` flag).

pokgak commented on examples/dtls-sock/tinydtls_keys.h:
```diff
@@ -0,0 +1,66 @@
+ * @{
+ *
+ * @file
+ * @brief       PSK and RPK keys for the dtls-echo example.
```
Changed `dtls-echo` to `dtls-sock`. No other changes are made to the file so I skipped the author and copyright.

pokgak commented on examples/dtls-sock/dtls-server.c:
```diff
@@ -0,0 +1,199 @@
+                                  10 * US_PER_SEC);
+            if (res < 0) {
+                if (res != -ETIMEDOUT) {
+                    printf("Error receiving UDP over DTLS %d", (int)res);
```
More detailed error messages can be enabled by setting the `ENABLE_DEBUG` flag in `sock_dtls.c` or `DTLS_DEBUG` flag in Makefile. This is just to signal to the user that there is an error. The user can also check the return code to see what error is returned specifically.

pokgak commented:
Glad to hear you liked it @rfuentess!

I rebased the code to latest master to pull in the fixes from #12478. The tests from https://github.com/RIOT-OS/RIOT/pull/11943#issuecomment-539625537 should now all passes but I haven't tested it myself yet. I'll try to find time for it next week.

rfuentess commented:
I’m going to be unavailable to test further the PR until mid-November. Please, feel free of discarding my review if the merge must happen before that date.

pokgak commented:
Noted @rfuentess .

@smlng ping! I addressed all your comments.

I rerun the tests after pulling the fixes from other issues in https://github.com/RIOT-OS/RIOT/pull/11943#issuecomment-539625537. The test is a simple packet send from the client to the server and see the output log. The results are as follows:

```
sock = dtls-sock
echo = dtls-echo

-------- -------------- -------- -------------- ---------------------
client   cipher suite   server   cipher suite   result
                                                ---------------------
                                                native | samr21-xpro
-------- -------------- -------- -------------- ---------------------
sock     ecc            sock     ecc            OK     | OK
sock     psk            sock     psk            OK     | OK
sock     ecc            sock     psk            OK 1*  | OK 1*
sock     psk            sock     ecc            OK 1*  | OK 1*

sock     ecc            echo     ecc            OK     | OK
sock     psk            echo     psk            OK     | OK
sock     ecc            echo     psk            OK 1*  | OK 1*
sock     psk            echo     ecc            OK 1*  | OK 1*

echo     ecc            sock     ecc            OK     | OK
echo     psk            sock     psk            OK     | OK
echo     ecc            sock     psk            OK 1*  | OK 1*
echo     psk            sock     ecc            OK 1*  | OK 1*
-------- -------------- -------- -------------- ---------------------
1*: client got handshake failure as expected
```

pokgak commented:
@smlng if you have the time, would you mind taking over the review?

smlng commented:
please fix cppcheck issues and squash afterwards into reasonable number of meaningful commits, i.e. pkg and example separately

pokgak commented:
cppcheck error fixed and squashed commits. I need to read https://github.com/RIOT-OS/RIOT/pull/11943/commits/b7d0355ed8bf20c7b88eaeac3f6b18f13c2d6246 (originally included in #11909) because it somehow got removed from master.

tcschmidt commented:
@smlng sind die requests erledigt? Dann bitte den Block aufheben.

smlng commented:
please squash

pokgak commented:
Squashed.

pokgak commented:
>     * cannot start, then stop and start server again, fails with:

Fixed.



>     * further it would be nice to print the message send by `dtlsc` on server side and/or on client side (after echo). Currently both only report that a message was received but the content is not shown.

I had this discussion with @miri64 before (https://github.com/RIOT-OS/RIOT/pull/11909#discussion_r312498087) and there decided against printing the message.

smlng commented:
please squash, otherwise ACK.

pokgak commented:
@smlng sorry for the delay.

I tested part of the tests from https://github.com/RIOT-OS/RIOT/pull/11943#issuecomment-546914277 again on `native` and all OK:

```
-------- -------------- -------- -------------- ---------------------
client   cipher suite   server   cipher suite   result
                                                ---------------------
                                                native 
-------- -------------- -------- -------------- ---------------------
sock     ecc            sock     ecc            OK     
sock     psk            sock     psk            OK    
sock     ecc            sock     psk            OK 1* 
sock     psk            sock     ecc            OK 1* 

1*: client got handshake failure as expected
```

smlng commented:
okay ... cool, lets see what murdock thinks

benpicco commented:
Murdock is happy and so am I 😃

smlng commented:
well then let's do this!

rfuentess commented:
Congrats @pokgak !

pokgak commented:
Thank you all for helping with the review :)

benpicco commented on pkg/tinydtls/include/sock_dtls_types.h:
```diff
@@ -0,0 +1,66 @@
+/**
+ * @brief Information about DTLS sock
+ */
+struct sock_dtls {
+    dtls_context_t *dtls_ctx;               /**< TinyDTLS context for sock */
+    sock_udp_t *udp_sock;                   /**< Underlying UDP sock to use */
```
What was the reason for not doing
```suggestion
    sock_udp_t udp_sock;                    /**< Underlying UDP sock to use */
    dtls_context_t *dtls_ctx;               /**< TinyDTLS context for sock */
```
instead?

That way application code using UDP sockets could transparently handle DTLS sockets (after a custom setup procedure).
`sock_udp_t` would only need to be extended by a flag to indicate there is a DTLS context attached to it.

miri64 replied:
We might want to support other transports in the future. With that, it would be easier and better understandable to create the transport sock in the application first and then the DTLS sock. As we do not want to copy that transport sock, a pointer is the best way to go here.
